Reader Comments: Add fallback implementation for overflow menu on iOS 13

dvdchr opened the pull request:
Refs #17476 

This PR adds `MenuSheetViewController`, which is used to mimic the appearance of `UIContextMenu` in iOS 13. The main difference is that this can be fully customizable – can be triggered on tap, or shown programmatically – unlike `UIContextMenu` which can only be shown on long press.

The component is displayed using a `UIPopoverPresentationController` with a "forced" setting on iPhone such that it always shows the menu with `.popover` presentation style (instead of `.formSheet`). 

Here's a preview:

• | `UIMenu`, iOS 14+ | Fallback implementation, iOS 13
--|--|--
Light scheme | ![menu_light](https://user-images.githubusercontent.com/1299411/144654838-f0c522b3-61cd-4216-aa04-2d225924da76.png) | ![fallback_light](https://user-images.githubusercontent.com/1299411/144654835-52e12da9-5af6-4a7f-9ab7-ae6eeab9adf9.png)
Dark scheme | ![menu_dark](https://user-images.githubusercontent.com/1299411/144654814-59ea4ece-c2ae-40da-8b1c-71f4dea64fc5.png) | ![fallback_dark](https://user-images.githubusercontent.com/1299411/144654828-6d381180-e18f-4366-be18-115241df8241.png)

There are slight differences visually (positioning, caret, visual effects, shadows, etc.), but I'm thinking this is good enough to be implemented in a short amount of time.

## To Test

Ensure that the `newCommentThread` feature flag is enabled.

#### With iOS 14+:

- Go to any comment thread on a site that you can moderate. (For example, My Site > Comments > tap any comment reply > tap on the header.)
- Tap on the ellipsis button.
- 🔍  Verify that the `UIMenu` is shown.
- Tap on the Share menu.
- 🔍  Verify that the share sheet is presented.

#### With iOS 13:

(Tip: if you don't have iOS 13 simulator downloaded, you can temporarily disable the code in ReaderCommentsViewController.swift, particularly the if #available lines.)

- Go to any comment thread on a site that you can moderate. (For example, My Site > Comments > tap any comment reply > tap on the header.)
- Tap on the ellipsis button.
- 🔍  Verify that the menu is shown.
- Tap on the Share menu.
- 🔍  Verify that the share sheet is presented.

## Regression Notes
1. Potential unintended areas of impact
n/a. Feature is unreleased.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a. Feature is unreleased.

3. What automated tests I added (or what prevented me from doing so)
n/a. Feature is unreleased.


PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Add MenuSheetViewController

dvdchr committed:
Update MenuSheetViewController to be displayed modally

dvdchr committed:
Improve documentation

dvdchr committed:
Use safe index method

ScoutHarris commented:
Hey @dvdchr . Got a few issues for ya.

**iOS13:**

Menu disappears on rotation:
- In portrait, show the menu on a comment close to the bottom of the list.
- Rotate to landscape.
- Notice the menu is not shown (I'm assuming because the comment isn't).
- The menu is shown again when rotating back to portrait.
- Rotate a few more times. Notice the app becomes unresponsive. It doesn't appear to crash, it just stops responding.

| ![missing_portrait](https://user-images.githubusercontent.com/1816888/144680904-107ae1e2-4f7a-4d0f-b6c2-f99ac82612e9.png) | ![missing_land](https://user-images.githubusercontent.com/1816888/144680896-b2249632-f457-41bd-9e65-d9037bb71e74.png) |
|--------|-------|


Menu is misplaced on rotation:
- In portrait, show the menu on a comment in the middle of the list.
- Rotate to landscape.
- Notice the menu is not inline with the comment button.

| ![misplaced_port](https://user-images.githubusercontent.com/1816888/144680983-5f873e9b-83d6-4470-b975-3b49cf429b7d.png) | ![misplaced_land](https://user-images.githubusercontent.com/1816888/144680976-4df97520-dfcd-4e5f-a9a6-ecc53582cd46.png) |
|--------|-------|

Menu UI:
- There appears to be an extra row on top of the menu.

<img width="348" alt="extra_row" src="https://user-images.githubusercontent.com/1816888/144681060-1b3137fc-178c-4054-9cf1-1a7007c49419.png">


**iOS14.5:**

Menu disappears on rotation:
- In portrait, show the menu on any comment.
- Rotate to landscape.
- Notice the menu is not shown.
- Rotating back to portrait doesn't make it reappear as it does in iOS13.

ScoutHarris reviewed (changes requested):
Hey @dvdchr . I know this is a WIP, but I'm requesting changes primarily due to the app becoming unresponsive in iOS13.

dvdchr committed:
Re-add section header height for MenuSheetViewController

dvdchr committed:
Use automatic dimension for estimated row height

Using the previous row height constant caused the app to hang due to some cells being constantly reused and configured.

dvdchr committed:
Dismiss menu when orientation changes

dvdchr commented:
Thanks for the details and screenshots @ScoutHarris ! 

> iOS 13:
> 
> Menu disappears on rotation:

When fixing this, I noticed that it will be a totally different approach with iOS 14.5 since it uses a `UIMenu`. However, after searching around, it looks like the system intentionally dismisses menus on orientation change – even when it changes from landscape to landscape. I've tried this by going to Apple's Messages > conversation screen in landscape mode. While in split view, I tapped on the ellipsis menu and it also shows a `UIMenu`. 

Therefore, I figured it's best for the fallback implementation to follow the `UIMenu` behavior instead. This has been addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17622/commits/d7ba1ff1e9fc6e10bceedc28c3d0506bb2e6f862. Note that this should also address these comments:

- iOS 13: Menu is misplaced on rotation -> Menu will now be dismissed on rotation.
- iOS 14.5: Menu disappears on rotation -> This is expected behavior.

> Menu UI:
>
> There appears to be an extra row on top of the menu.

Thank you for noticing this! I've initially fixed this, but somehow it got removed in [this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/17622/commits/ec95d71186ddb873f2e4d8768ceea04ebf6fe7c8#diff-ef2b4c45e4793d7105ec1bd0ec1835763e183a7fef4b0fc56eaf7328109a9d7cL114). I've re-added this in https://github.com/wordpress-mobile/WordPress-iOS/pull/17622/commits/1eb9d2434e70e03c3e0f6a1227835b5d656a8282. 🙂 

> Rotate a few more times. Notice the app becomes unresponsive. It doesn't appear to crash, it just stops responding.

Now, this is an interesting one. After investigating this, it's actually unrelated to the menu and it also happens in iOS 14+. The  problem is, the `ReaderCommentsViewController` has a constant, `EstimatedCommentRowHeight`, which is used in the `estimatedHeightForRow` method. When scrolling through the new comment threads when it has a lot of text per comments, the `prepareForReuse` and `configureCell` would get stuck in an infinite loop and this caused the app to be unresponsive 😬 .

In  https://github.com/wordpress-mobile/WordPress-iOS/pull/17622/commits/5887a6106ce9f689d601d132734041ebc38683aa, I've updated the `estimatedHeightForRow` to use automatic dimension addresses this issue. The thread still jumps around as you scroll, but I think this is a lot better than turning unresponsive.

Ready for another round of review!

dvdchr committed:
Add self in closure

ScoutHarris commented:
> iOS 13: Menu is misplaced on rotation -> Menu will now be dismissed on rotation.
> iOS 14.5: Menu disappears on rotation -> This is expected behavior.

Sounds good. ➕ 1️⃣ for consistency. 😄 

> The thread still jumps around as you scroll, but I think this is a lot better than turning unresponsive.

Agreed!

dvdchr commented:
Thanks @ScoutHarris !
